module5/lessons/Week1/ErrorHandling.md

```diff
@@ -0,0 +1,326 @@
+        public List<Review> Reviews { get; set; } = new List<Review>();
+    }
+```
+
+### Data Validation Step 2: The Controller
+Now we need to add a check to see if the incoming data meets these validations in our `MoviesController.cs`.
+
+```c#
+[HttpPost]
+public IActionResult Index(Movie movie)
+{
+    //ModelState.IsValid is a built in .NET tool we can use that will check all of the validations on our models and then attach any error messages to the object. 
+    if (ModelState.IsValid) // This is new
+    {
+        //Take the movie sent in the request and save it to the database
+        _context.Movies.Add(movie);
+        _context.SaveChanges();
+
+        // The id generated by the database is now on the object we added to the context
+        var newMovieId = movie.Id;
+
+        // Redirect to our route /movies/show and pass in the newMovieId for the id parameter
+        return RedirectToAction("show", new { id = newMovieId });
+    }
+    else // This is new
+    {
+        // WHAT DO WE WANT TO DO IF VALIDATION FAILS?
+    }
+}
+```
+
+<section class="call-to-action" markdown="1">
+With your partner: Discuss what you think we want to happen if the validation fails on the movie passed in.
+</section>
+
+<section class="answer" markdown="1">
+### What we want to do
+
+We want to send the user to the New page and pass in the error information to display for the user. The error info is all on the model so we will pass the model to the view.
+```c#
+//Render the New view and pass in the movie so that we have the validation information
+return View("New", movie);
+```
+</section>
+
+
+### Data Validation Step 3: The View
+We need to allow a movie to get passed into our view and display any available validation messages.
+
+```html
+@model Movie <----This is new
+
+<h1>Add a Movie</h1>
+
+<form method="post" action="/movies">
+    <div class="form-group">
+        <label for="Title">Title:</label>
+        <input type="text" id="Title" name="Title" />
+        <span asp-validation-for="Title" class="text-danger"></span> <--- This is new
+    </div>
+    <div class="form-group">
+        <label for="Genre">Genre:</label>
+        <input type="text" id="Genre" name="Genre" />
+        <span asp-validation-for="Genre" class="text-danger"></span>  <--- This is new
+    </div>
+    <button type="submit">Create Movie</button>
+</form>
+```
+
+<aside class="instructor-notes" markdown="1">
+    <p><strong>Instructor Note</strong><br>
+    This code will hook up to the model and display the current value if there is one.
+    <label asp-for="Title">Title:</label>
+    <input asp-for="Title" class="form-control" />
+    </p>
+</aside>
+
+It's out of scope of this lesson, but if you want to display what the user previous input I recommend looking into 
+.NET tag helpers, specifically [asp-for](https://learn.microsoft.com/en-us/aspnet/core/mvc/views/working-with-forms?view=aspnetcore-5.0#the-input-tag-helper).
+

```

Comment:
I think we should call out the difference in the server-side validations, and client side validations here.  Something like 'using the asp-validation-for send instructions to the form to prevent 'bad' data from being submitted`.  We could re-inforce the client-side nature by inspecting this form to see the raw html that is controlling these validations.